File: stepper_test.c
#include <gpiod.h>
#include <unistd.h>
#include <stdio.h>

#define CHIP_NAME "/dev/gpiochip1"
#define STEP_LINE_OFFSET 25  // PIN_18
#define DIR_LINE_OFFSET 24   // PIN_17

int main(void) {
    struct gpiod_chip *chip;
    struct gpiod_line *step, *dir;

    chip = gpiod_chip_open(CHIP_NAME);
    if (!chip) {
        perror("Failed to open GPIO chip");
        return 1;
    }

    step = gpiod_chip_get_line(chip, STEP_LINE_OFFSET);
    dir = gpiod_chip_get_line(chip, DIR_LINE_OFFSET);

    if (!step || !dir) {
        perror("Failed to get GPIO lines");
        return 1;
    }

    if (gpiod_line_request_output(step, "stepper", 0) < 0 ||
        gpiod_line_request_output(dir, "stepper", 0) < 0) {
        perror("Failed to request lines as output");
        return 1;
    }
// Set motor direction
    gpiod_line_set_value(dir, 1);  // 1 = CW, 0 = CCW

    // Generate step pulses
    for (int i = 0; i < 200; i++) {  // Adjust for full revs or microstepping
        gpiod_line_set_value(step, 1);
        usleep(1000);  // 1 ms HIGH
        gpiod_line_set_value(step, 0);
        usleep(1000);  // 1 ms LOW
    }

    gpiod_line_release(step);
    gpiod_line_release(dir);
    gpiod_chip_close(chip);

    return 0;
}

File: stepper_test.py
import gpiod
import time

# GPIO chip and lines
CHIP = 'gpiochip1'
STEP_LINE = 25  # PIN_18
DIR_LINE = 24   # PIN_17

# Setup chip and lines
chip = gpiod.Chip(CHIP)
step = chip.get_line(STEP_LINE)
direction = chip.get_line(DIR_LINE)

# Request lines as outputs
step.request(consumer='stepper', type=gpiod.LINE_REQ_DIR_OUT)
direction.request(consumer='stepper', type=gpiod.LINE_REQ_DIR_OUT)

# Set direction: 1 = CW, 0 = CCW
direction.set_value(1)

# Send steps
num_steps = 200  # One full revolution for 1.8° stepper motor
delay = 0.001    # 1 ms = 1 kHz → adjust for speed

for i in range(num_steps):
    step.set_value(1)
    time.sleep(delay)
    step.set_value(0)
    time.sleep(delay)

# Cleanup
step.set_value(0)
direction.set_value(0)
step.release()
direction.release()
